test/test_asarray.py

```diff
@@ -0,0 +1,423 @@
+
+        # 1. The storage pointers should be equal only if 'is_alias' is set
+        if is_alias:
+            self.assertEqual(result.data_ptr(), getaddr(original))
+        else:
+            self.assertNotEqual(result.data_ptr(), getaddr(original))
+
+        # 2. Comparison of the elements only takes place if the original
+        # sequence and the resulting tensor have the same data type
+        if same_dtype:
+            self.assertEqual(original_tensor, result)
+        else:
+            dtype = kwargs.get("dtype", torch.get_default_dtype())
+            self.assertEqual(original_tensor.shape, result.shape)
+            self.assertEqual(dtype, result.dtype)
+
+
+        # 3. Given the specified target device, we first check whether
+        # its type is the same, and then if its index is the same (if it
+        # is not None)
+        if same_device:
+            device = getdevice(original)
+        else:
+            device = torch.device(kwargs.get("device", "cpu"))
+
+        # Compare the target device type, and its index
+        self.assertEqual(device.type, result.device.type)
+        if device.index is not None:
+            self.assertEqual(device.index, result.device.index)
+
+        # 4. By default, 'requires_grad' is unset
+        self.assertEqual(result.requires_grad, kwargs.get("requires_grad", False))
+
+    def _may_compute_grad(self, dtype):
+        return dtype.is_floating_point or dtype.is_complex
+
+    # Skipping 'meta' devices, since there's no point in comparing their
+    # data pointer (which is basically the point here), since they all
+    # return 0.
+    @skipMeta
+    @dtypes(*get_all_dtypes())
+    def test_alias_from_tensor(self, device, dtype):
+        # DLpack does not explicitly support bool
+        # It does it through uint8 type
+        if dtype is torch.bool:
+            return
+
+        original = common.make_tensor((5, 5), device, dtype)
+
+        def check(**kwargs):
+            self._check(original, **kwargs)
+
+        check(requires_grad=self._may_compute_grad(dtype))
+        check(copy=False)
+        check(dtype=dtype)
+        check(dtype=dtype, copy=False)
+        check(device=device)
+        check(device=device, dtype=dtype)
+        check(device=device, dtype=dtype, copy=False)
+
+    @skipMeta
+    @dtypes(*get_all_dtypes())
+    def test_copy_tensor(self, device, dtype):
+        if dtype is torch.bool:
+            return
+
+        original = common.make_tensor((5, 5), device, dtype)
+
+        def check(**kwargs):
+            self._check(original, is_alias=False, **kwargs)
+
+        check(requires_grad=self._may_compute_grad(dtype), copy=True)
+        check(dtype=dtype, copy=True)
+        check(device=device, copy=True)
+        check(device=device, dtype=dtype, copy=True)
+
+        # Copy is forced because of different device
+        if torch.cuda.is_available():
+            other = "cuda" if device == "cpu" else "cpu"
```

Comment:
Assert an error here if `copy=False`

Reply:
Don't compare device to a string, instead use `self.device_type` or `torch.device(device).type`